oxidation/crates/parsec_api_protocol/src/handshake.rs

```diff
@@ -0,0 +1,653 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+use std::cmp::Ordering;
+
+use parsec_api_crypto::{SigningKey, VerifyKey};
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, Bytes};
+
+use crate::{impl_dumps_loads, ChallengeDataReport, HandshakeError, InvitationType};
+use parsec_api_types::{maybe_field, DateTime, DeviceID, InvitationToken, OrganizationID};
+
+const HANDSHAKE_CHALLENGE_SIZE: usize = 48;
+
+pub const API_V1_VERSION: ApiVersion = ApiVersion {
+    version: 1,
+    revision: 3,
+};
+pub const API_V2_VERSION: ApiVersion = ApiVersion {
+    version: 2,
+    revision: 5,
+};
+pub const API_VERSION: ApiVersion = API_V2_VERSION;
+pub const BALLPARK_CLIENT_EARLY_OFFSET: f64 = 50.0; // seconds
+pub const BALLPARK_CLIENT_LATE_OFFSET: f64 = 70.0; // seconds
+const BALLPARK_CLIENT_TOLERANCE: f64 = 0.8; // 80%
+
+pub fn timestamps_in_the_ballpark(
+    client: DateTime,
+    backend: DateTime,
+    ballpark_client_early_offset: f64,
+    ballpark_client_late_offset: f64,
+) -> bool {
+    // Useful to compare signed message timestamp with the one stored by the
+    // backend.
+    let seconds = (backend - client).num_seconds() as f64;
+    -ballpark_client_early_offset < seconds && seconds < ballpark_client_late_offset
+}
+
+#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ApiVersion {
+    pub version: u32,
+    pub revision: u32,
+}
+
+impl PartialOrd for ApiVersion {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ApiVersion {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.version.cmp(&other.version) {
+            Ordering::Equal => self.revision.cmp(&other.revision),
+            order => order,
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(tag = "type")]
+pub enum Answer {
+    #[serde(rename = "AUTHENTICATED")]
+    Authenticated {
+        client_api_version: ApiVersion,
+        organization_id: OrganizationID,
+        device_id: DeviceID,
+        rvk: Box<VerifyKey>,
+        #[serde_as(as = "Bytes")]
+        answer: Vec<u8>,
+    },
+    #[serde(rename = "INVITED")]
+    Invited {
+        client_api_version: ApiVersion,
+        organization_id: OrganizationID,
+        invitation_type: InvitationType,
+        token: InvitationToken,
+    },
+    #[serde(rename = "signed_answer")]
```

Comment:
`AnswerSchema` is different from `HandshakeAuthenticatedAnswerSchema`/`HandshakeInvitedAnswerSchema` !

`AnswerSchema` is use to build the `HandshakeAuthenticatedAnswerSchema.answer` field (by serializing then signing)

Reply:
Done :+1: